cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/DistributedStreamService.java

```diff
@@ -381,11 +378,27 @@ private void performLeaderElection() {
       zkClient, "/election/" + STREAMS_COORDINATOR, new ElectionHandler() {
       @Override
       public void leader() {
-        LOG.info("Became Stream handler leader. Starting resource coordinator.");
-        resourceCoordinator = new ResourceCoordinator(getCoordinatorZKClient(), discoveryServiceClient,
-                                                      new BalancedAssignmentStrategy());
-        resourceCoordinator.startAndWait();
-        updateRequirement();
+        while (true) {
+          try {
+            LOG.info("Became Stream handler leader. Starting resource coordinator.");
+            heartbeatsSubscriptionExecutor = Executors.newSingleThreadExecutor(
+              Threads.createDaemonThreadFactory("heartbeats-subscription-executor"));
+            heartbeatsSubscription = subscribeToHeartbeatsFeed();
+            resourceCoordinator = new ResourceCoordinator(getCoordinatorZKClient(), discoveryServiceClient,
+                                                          new BalancedAssignmentStrategy());
+            resourceCoordinator.startAndWait();
```

Comment:
So you'll have a lot of resource coordinator started if updateRequirement throws exception?